ccl/streamingccl/streamingest: TestRandomClientGeneration failed

ccl/streamingccl/streamingest.TestRandomClientGeneration [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11534127?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11534127?buildTab=artifacts#/) on master @ [85cbfffeaa60c3d40e51eb6be7e49eca4dcc8a18](https://github.com/cockroachdb/cockroach/commits/85cbfffeaa60c3d40e51eb6be7e49eca4dcc8a18):


```
=== RUN   TestRandomClientGeneration
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/90c1b75d835f45b8488807abb5b1092d/logTestRandomClientGeneration1846648453
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:131: automatically injected virtual cluster under test; see comment at top of test_server_shim.go for details.
    stream_ingestion_processor_test.go:522: Partition Span /Table/500{0-1}; Partition ID 5000
    stream_ingestion_processor_test.go:522: Partition Span /Table/500{1-2}; Partition ID 5001
    stream_ingestion_processor_test.go:539: target span /Tenant/30/Table/500{1-2}; partition ts 1693297911.398104189,0
    stream_ingestion_processor_test.go:539: target span /Tenant/30/Table/500{0-1}; partition ts 1693297911.372978559,0
    stream_ingestion_processor_test.go:747: -- test log scope end --
    stream_ingestion_processor_test.go:747: Leaked goroutine: goroutine 472456 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).checkForCutoverSignal(0xc00e4a5400, {0xd7c61b8, 0xc007f1dcc0})
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:603 +0x249
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).Start.func2({0xd7c61b8, 0xc007f1dcc0})
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:481 +0x45
        github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1()
        	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168 +0x4d
        golang.org/x/sync/errgroup.(*Group).Go.func1()
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x87
        created by golang.org/x/sync/errgroup.(*Group).Go
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:72 +0x12f
        Leaked goroutine: goroutine 472457 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).flush(0xc00e4a5400)
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:1165 +0x2f4
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).consumeEvents(0xc00e4a5400, {0xc011821da0?, 0xc011821e00?})
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:674 +0x2af
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).Start.func3({0xd7c61b8, 0xc007f1dcc0})
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:488 +0xc5
        github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1()
        	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168 +0x4d
        golang.org/x/sync/errgroup.(*Group).Go.func1()
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x87
        created by golang.org/x/sync/errgroup.(*Group).Go
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:72 +0x12f
--- FAIL: TestRandomClientGeneration (47.19s)
```
<p>Parameters: <code>TAGS=bazel,gss,race</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/disaster-recovery
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomClientGeneration.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-31037

ccl/streamingccl/streamingest.TestRandomClientGeneration [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11534065?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11534065?buildTab=artifacts#/) on master @ [85cbfffeaa60c3d40e51eb6be7e49eca4dcc8a18](https://github.com/cockroachdb/cockroach/commits/85cbfffeaa60c3d40e51eb6be7e49eca4dcc8a18):


```
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/merged_subscription.go:68
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).Start.func1({0xc00953bf20?, 0xc010e69440?})
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:475 +0x39
        github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1()
        	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168 +0x25
        golang.org/x/sync/errgroup.(*Group).Go.func1()
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x64
        created by golang.org/x/sync/errgroup.(*Group).Go
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:72 +0xa5
        Leaked goroutine: goroutine 173206854 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).checkForCutoverSignal(0xc016561800, {0x7a7a518, 0xc00fa85d80})
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:603 +0x10f
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).Start.func2({0x7a7a518?, 0xc00fa85d80?})
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:481 +0x35
        github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1()
        	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168 +0x25
        golang.org/x/sync/errgroup.(*Group).Go.func1()
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x64
        created by golang.org/x/sync/errgroup.(*Group).Go
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:72 +0xa5
        Leaked goroutine: goroutine 173206855 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).flush(0xc016561800)
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:1165 +0x19d
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).consumeEvents(0xc016561800, {0xc010cf79f0?, 0x8?})
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:696 +0x125
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).Start.func3({0x7a7a518?, 0xc00fa85d80?})
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:488 +0x7d
        github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1()
        	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168 +0x25
        golang.org/x/sync/errgroup.(*Group).Go.func1()
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x64
        created by golang.org/x/sync/errgroup.(*Group).Go
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:72 +0xa5
        Leaked goroutine: goroutine 173206856 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).flushLoop(0xc016561800, {0xc010cf79a0?, 0x0?})
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:648 +0xd1
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).Start.func4({0x7a7a518?, 0xc00fa85d80?})
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:495 +0x7d
        github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1()
        	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168 +0x25
        golang.org/x/sync/errgroup.(*Group).Go.func1()
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x64
        created by golang.org/x/sync/errgroup.(*Group).Go
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:72 +0xa5
        Leaked goroutine: goroutine 175727467 [chan receive]:
        github.com/cockroachdb/pebble/sstable.(*writeQueue).runWorker(0xc01d239c00)
        	github.com/cockroachdb/pebble/sstable/external/com_github_cockroachdb_pebble/sstable/write_queue.go:98 +0x4c
        created by github.com/cockroachdb/pebble/sstable.newWriteQueue
        	github.com/cockroachdb/pebble/sstable/external/com_github_cockroachdb_pebble/sstable/write_queue.go:57 +0xd6
--- FAIL: TestRandomClientGeneration (29.40s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomClientGeneration.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamingest.TestRandomClientGeneration [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11549635?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11549635?buildTab=artifacts#/) on master @ [36bd8296bad475a8183e474bf4aeb08bb83f8a2d](https://github.com/cockroachdb/cockroach/commits/36bd8296bad475a8183e474bf4aeb08bb83f8a2d):


```
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/merged_subscription.go:68
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).Start.func1({0xc01d493560?, 0xc015f3cde0?})
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:475 +0x39
        github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1()
        	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168 +0x25
        golang.org/x/sync/errgroup.(*Group).Go.func1()
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x64
        created by golang.org/x/sync/errgroup.(*Group).Go
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:72 +0xa5
        Leaked goroutine: goroutine 171510087 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).checkForCutoverSignal(0xc002a67000, {0x7a88d38, 0xc001022980})
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:603 +0x10f
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).Start.func2({0x7a88d38?, 0xc001022980?})
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:481 +0x35
        github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1()
        	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168 +0x25
        golang.org/x/sync/errgroup.(*Group).Go.func1()
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x64
        created by golang.org/x/sync/errgroup.(*Group).Go
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:72 +0xa5
        Leaked goroutine: goroutine 171510088 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).flush(0xc002a67000)
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:1165 +0x19d
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).consumeEvents(0xc002a67000, {0xc01f674820?, 0x1b47f51e92c09e78?})
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:696 +0x125
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).Start.func3({0x7a88d38?, 0xc001022980?})
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:488 +0x7d
        github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1()
        	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168 +0x25
        golang.org/x/sync/errgroup.(*Group).Go.func1()
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x64
        created by golang.org/x/sync/errgroup.(*Group).Go
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:72 +0xa5
        Leaked goroutine: goroutine 171510089 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).flushLoop(0xc002a67000, {0xc01f6747d0?, 0x17801c469ecd9b64?})
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:648 +0xd1
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).Start.func4({0x7a88d38?, 0xc001022980?})
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:495 +0x7d
        github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1()
        	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168 +0x25
        golang.org/x/sync/errgroup.(*Group).Go.func1()
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x64
        created by golang.org/x/sync/errgroup.(*Group).Go
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:72 +0xa5
        Leaked goroutine: goroutine 174273809 [chan receive]:
        github.com/cockroachdb/pebble/sstable.(*writeQueue).runWorker(0xc017fb3200)
        	github.com/cockroachdb/pebble/sstable/external/com_github_cockroachdb_pebble/sstable/write_queue.go:98 +0x4c
        created by github.com/cockroachdb/pebble/sstable.newWriteQueue
        	github.com/cockroachdb/pebble/sstable/external/com_github_cockroachdb_pebble/sstable/write_queue.go:57 +0xd6
--- FAIL: TestRandomClientGeneration (36.10s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomClientGeneration.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamingest.TestRandomClientGeneration [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11565965?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11565965?buildTab=artifacts#/) on master @ [6f589302865e24a3843a31cd7686514b5430f626](https://github.com/cockroachdb/cockroach/commits/6f589302865e24a3843a31cd7686514b5430f626):


```
=== RUN   TestRandomClientGeneration
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/90c1b75d835f45b8488807abb5b1092d/logTestRandomClientGeneration2146321691
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:132: automatically injected virtual cluster under test; see comment at top of test_server_shim.go for details.
    stream_ingestion_processor_test.go:522: Partition Span /Table/500{0-1}; Partition ID 5000
    stream_ingestion_processor_test.go:522: Partition Span /Table/500{1-2}; Partition ID 5001
    stream_ingestion_processor_test.go:539: target span /Tenant/30/Table/500{1-2}; partition ts 1693470419.871072242,0
    stream_ingestion_processor_test.go:539: target span /Tenant/30/Table/500{0-1}; partition ts 1693470419.860337066,0
    stream_ingestion_processor_test.go:747: -- test log scope end --
    stream_ingestion_processor_test.go:747: Leaked goroutine: goroutine 454649 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).checkForCutoverSignal(0xc013094400, {0xd7ea4b8, 0xc0089808c0})
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:603 +0x249
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).Start.func2({0xd7ea4b8, 0xc0089808c0})
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:481 +0x45
        github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1()
        	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168 +0x4d
        golang.org/x/sync/errgroup.(*Group).Go.func1()
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x87
        created by golang.org/x/sync/errgroup.(*Group).Go
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:72 +0x12f
        Leaked goroutine: goroutine 454650 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).flush(0xc013094400)
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:1165 +0x2f4
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).consumeEvents(0xc013094400, {0x0?, 0x0?})
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:674 +0x2af
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).Start.func3({0xd7ea4b8, 0xc0089808c0})
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:488 +0xc5
        github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1()
        	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168 +0x4d
        golang.org/x/sync/errgroup.(*Group).Go.func1()
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x87
        created by golang.org/x/sync/errgroup.(*Group).Go
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:72 +0x12f
--- FAIL: TestRandomClientGeneration (44.15s)
```
<p>Parameters: <code>TAGS=bazel,gss,race</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomClientGeneration.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamingest.TestRandomClientGeneration [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11565911?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11565911?buildTab=artifacts#/) on master @ [6f589302865e24a3843a31cd7686514b5430f626](https://github.com/cockroachdb/cockroach/commits/6f589302865e24a3843a31cd7686514b5430f626):


```
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/merged_subscription.go:68
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).Start.func1({0x69620e0a33420001?, 0xc010cd2ae0?})
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:475 +0x39
        github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1()
        	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168 +0x25
        golang.org/x/sync/errgroup.(*Group).Go.func1()
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x64
        created by golang.org/x/sync/errgroup.(*Group).Go
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:72 +0xa5
        Leaked goroutine: goroutine 178557505 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).checkForCutoverSignal(0xc010c66800, {0x7a96498, 0xc007e8f7c0})
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:603 +0x10f
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).Start.func2({0x7a96498?, 0xc007e8f7c0?})
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:481 +0x35
        github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1()
        	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168 +0x25
        golang.org/x/sync/errgroup.(*Group).Go.func1()
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x64
        created by golang.org/x/sync/errgroup.(*Group).Go
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:72 +0xa5
        Leaked goroutine: goroutine 178557522 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).flush(0xc010c66800)
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:1165 +0x19d
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).consumeEvents(0xc010c66800, {0xc01e9e05a0?, 0xb?})
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:696 +0x125
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).Start.func3({0x7a96498?, 0xc007e8f7c0?})
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:488 +0x7d
        github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1()
        	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168 +0x25
        golang.org/x/sync/errgroup.(*Group).Go.func1()
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x64
        created by golang.org/x/sync/errgroup.(*Group).Go
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:72 +0xa5
        Leaked goroutine: goroutine 178557523 [select]:
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).flushLoop(0xc010c66800, {0xc01e9e0550?, 0x17806af517d746ba?})
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:648 +0xd1
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.(*streamIngestionProcessor).Start.func4({0x7a96498?, 0xc007e8f7c0?})
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:495 +0x7d
        github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1()
        	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168 +0x25
        golang.org/x/sync/errgroup.(*Group).Go.func1()
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x64
        created by golang.org/x/sync/errgroup.(*Group).Go
        	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:72 +0xa5
        Leaked goroutine: goroutine 181010793 [chan receive]:
        github.com/cockroachdb/pebble/sstable.(*writeQueue).runWorker(0xc019ab17c0)
        	github.com/cockroachdb/pebble/sstable/external/com_github_cockroachdb_pebble/sstable/write_queue.go:98 +0x4c
        created by github.com/cockroachdb/pebble/sstable.newWriteQueue
        	github.com/cockroachdb/pebble/sstable/external/com_github_cockroachdb_pebble/sstable/write_queue.go:57 +0xd6
--- FAIL: TestRandomClientGeneration (36.48s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomClientGeneration.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/backupccl/backup_processor.go
+++ b/pkg/ccl/backupccl/backup_processor.go
@@ -38,7 +38,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/mon"
-	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
 	"github.com/cockroachdb/cockroach/pkg/util/stop"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/cockroach/pkg/util/tracing"
@@ -173,11 +172,12 @@ func newBackupDataProcessor(
 			InputsToDrain: nil,
 			TrailingMetaCallback: func() []execinfrapb.ProducerMetadata {
 				bp.close()
-				meta := bp.constructTracingAggregatorProducerMeta(ctx)
-				if meta == nil {
-					return nil
+				if bp.agg != nil {
+					meta := bulk.ConstructTracingAggregatorProducerMeta(ctx,
+						bp.flowCtx.NodeID.SQLInstanceID(), bp.flowCtx.ID, bp.agg)
+					return []execinfrapb.ProducerMetadata{*meta}
 				}
-				return []execinfrapb.ProducerMetadata{*meta}
+				return nil
 			},
 		}); err != nil {
 		return nil, err
@@ -188,15 +188,17 @@ func newBackupDataProcessor(
 // Start is part of the RowSource interface.
 func (bp *backupDataProcessor) Start(ctx context.Context) {
 	ctx = logtags.AddTag(ctx, "job", bp.spec.JobID)
-	ctx = bp.StartInternal(ctx, backupProcessorName)
-	ctx, cancel := context.WithCancel(ctx)
 
 	// Construct an Aggregator to aggregate and render AggregatorEvents emitted in
 	// bps' trace recording.
-	ctx, bp.agg = bulk.MakeTracingAggregatorWithSpan(ctx,
-		fmt.Sprintf("%s-aggregator", backupProcessorName), bp.EvalCtx.Tracer)
+	bp.agg = bulk.TracingAggregatorForContext(ctx)
 	bp.aggTimer = timeutil.NewTimer()
-	bp.aggTimer.Reset(15 * time.Second)
+	// If the aggregator is nil, we do not want the timer to fire.
+	if bp.agg != nil {
+		bp.aggTimer.Reset(15 * time.Second)
+	}
+	ctx = bp.StartInternal(ctx, backupProcessorName, bp.agg)
+	ctx, cancel := context.WithCancel(ctx)
 
 	bp.cancelAndWaitForWorker = func() {
 		cancel()
@@ -246,36 +248,6 @@ func (bp *backupDataProcessor) constructProgressProducerMeta(
 	return &execinfrapb.ProducerMetadata{BulkProcessorProgress: &p}
 }
 
-func (bp *backupDataProcessor) constructTracingAggregatorProducerMeta(
-	ctx context.Context,
-) *execinfrapb.ProducerMetadata {
-	if bp.agg == nil {
-		return nil
-	}
-	aggEvents := &execinfrapb.TracingAggregatorEvents{
-		SQLInstanceID: bp.flowCtx.NodeID.SQLInstanceID(),
-		FlowID:        bp.flowCtx.ID,
-		Events:        make(map[string][]byte),
-	}
-	bp.agg.ForEachAggregatedEvent(func(name string, event bulk.TracingAggregatorEvent) {
-		msg, ok := event.(protoutil.Message)
-		if !ok {
-			// This should never happen but if it does skip the aggregated event.
-			log.Warningf(ctx, "event is not a protoutil.Message: %T", event)
-			return
-		}
-		data := make([]byte, msg.Size())
-		if _, err := msg.MarshalTo(data); err != nil {
-			// This should never happen but if it does skip the aggregated event.
-			log.Warningf(ctx, "failed to unmarshal aggregated event: %v", err.Error())
-			return
-		}
-		aggEvents.Events[name] = data
-	})
-
-	return &execinfrapb.ProducerMetadata{AggregatorEvents: aggEvents}
-}
-
 // Next is part of the RowSource interface.
 func (bp *backupDataProcessor) Next() (rowenc.EncDatumRow, *execinfrapb.ProducerMetadata) {
 	if bp.State != execinfra.StateRunning {
@@ -292,16 +264,14 @@ func (bp *backupDataProcessor) Next() (rowenc.EncDatumRow, *execinfrapb.Producer
 	case <-bp.aggTimer.C:
 		bp.aggTimer.Read = true
 		bp.aggTimer.Reset(15 * time.Second)
-		return nil, bp.constructTracingAggregatorProducerMeta(bp.Ctx())
+		return nil, bulk.ConstructTracingAggregatorProducerMeta(bp.Ctx(),
+			bp.flowCtx.NodeID.SQLInstanceID(), bp.flowCtx.ID, bp.agg)
 	}
 }
 
 func (bp *backupDataProcessor) close() {
 	bp.cancelAndWaitForWorker()
 	if bp.InternalClose() {
-		if bp.agg != nil {
-			bp.agg.Close()
-		}
 		bp.aggTimer.Stop()
 		bp.memAcc.Close(bp.Ctx())
 	}

--- a/pkg/ccl/backupccl/restore_data_processor.go
+++ b/pkg/ccl/backupccl/restore_data_processor.go
@@ -211,11 +211,12 @@ func newRestoreDataProcessor(
 			InputsToDrain: []execinfra.RowSource{input},
 			TrailingMetaCallback: func() []execinfrapb.ProducerMetadata {
 				rd.ConsumerClosed()
-				meta := rd.constructTracingAggregatorProducerMeta(ctx)
-				if meta == nil {
-					return nil
+				if rd.agg != nil {
+					meta := bulkutil.ConstructTracingAggregatorProducerMeta(ctx,
+						rd.flowCtx.NodeID.SQLInstanceID(), rd.flowCtx.ID, rd.agg)
+					return []execinfrapb.ProducerMetadata{*meta}
 				}
-				return []execinfrapb.ProducerMetadata{*meta}
+				return nil
 			},
 		}); err != nil {
 		return nil, err
@@ -226,14 +227,17 @@ func newRestoreDataProcessor(
 // Start is part of the RowSource interface.
 func (rd *restoreDataProcessor) Start(ctx context.Context) {
 	ctx = logtags.AddTag(ctx, "job", rd.spec.JobID)
-	ctx = rd.StartInternal(ctx, restoreDataProcName)
+	rd.agg = bulkutil.TracingAggregatorForContext(ctx)
+	rd.aggTimer = timeutil.NewTimer()
+	// If the aggregator is nil, we do not want the timer to fire.
+	if rd.agg != nil {
+		rd.aggTimer.Reset(15 * time.Second)
+	}
+
+	ctx = rd.StartInternal(ctx, restoreDataProcName, rd.agg)
 	rd.input.Start(ctx)
 
 	ctx, cancel := context.WithCancel(ctx)
-	ctx, rd.agg = bulkutil.MakeTracingAggregatorWithSpan(ctx, fmt.Sprintf("%s-aggregator", restoreDataProcName), rd.EvalCtx.Tracer)
-	rd.aggTimer = timeutil.NewTimer()
-	rd.aggTimer.Reset(15 * time.Second)
-
 	rd.cancelWorkersAndWait = func() {
 		cancel()
 		_ = rd.phaseGroup.Wait()
@@ -687,31 +691,6 @@ func makeProgressUpdate(
 	return progDetails
 }
 
-func (rd *restoreDataProcessor) constructTracingAggregatorProducerMeta(
-	ctx context.Context,
-) *execinfrapb.ProducerMetadata {
-	if rd.agg == nil {
-		return nil
-	}
-	aggEvents := &execinfrapb.TracingAggregatorEvents{
-		SQLInstanceID: rd.flowCtx.NodeID.SQLInstanceID(),
-		FlowID:        rd.flowCtx.ID,
-		Events:        make(map[string][]byte),
-	}
-	rd.agg.ForEachAggregatedEvent(func(name string, event bulkutil.TracingAggregatorEvent) {
-		var data []byte
-		var err error
-		if data, err = bulkutil.TracingAggregatorEventToBytes(ctx, event); err != nil {
-			// This should never happen but if it does skip the aggregated event.
-			log.Warningf(ctx, "failed to unmarshal aggregated event: %v", err.Error())
-			return
-		}
-		aggEvents.Events[name] = data
-	})
-
-	return &execinfrapb.ProducerMetadata{AggregatorEvents: aggEvents}
-}
-
 // Next is part of the RowSource interface.
 func (rd *restoreDataProcessor) Next() (rowenc.EncDatumRow, *execinfrapb.ProducerMetadata) {
 	if rd.State != execinfra.StateRunning {
@@ -738,7 +717,8 @@ func (rd *restoreDataProcessor) Next() (rowenc.EncDatumRow, *execinfrapb.Produce
 	case <-rd.aggTimer.C:
 		rd.aggTimer.Read = true
 		rd.aggTimer.Reset(15 * time.Second)
-		return nil, rd.constructTracingAggregatorProducerMeta(rd.Ctx())
+		return nil, bulkutil.ConstructTracingAggregatorProducerMeta(rd.Ctx(),
+			rd.flowCtx.NodeID.SQLInstanceID(), rd.flowCtx.ID, rd.agg)
 	case meta := <-rd.metaCh:
 		return nil, meta
 	case <-rd.Ctx().Done():
@@ -755,9 +735,6 @@ func (rd *restoreDataProcessor) ConsumerClosed() {
 	rd.cancelWorkersAndWait()
 
 	rd.qp.Close(rd.Ctx())
-	if rd.agg != nil {
-		rd.agg.Close()
-	}
 	rd.aggTimer.Stop()
 	rd.InternalClose()
 }

--- a/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go
+++ b/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go
@@ -10,7 +10,6 @@ package streamingest
 
 import (
 	"context"
-	"fmt"
 	"sort"
 	"sync"
 	"time"
@@ -330,11 +329,12 @@ func newStreamIngestionDataProcessor(
 			InputsToDrain: []execinfra.RowSource{},
 			TrailingMetaCallback: func() []execinfrapb.ProducerMetadata {
 				sip.close()
-				meta := sip.constructTracingAggregatorProducerMeta(ctx)
-				if meta == nil {
-					return nil
+				if sip.agg != nil {
+					meta := bulkutil.ConstructTracingAggregatorProducerMeta(ctx,
+						sip.flowCtx.NodeID.SQLInstanceID(), sip.flowCtx.ID, sip.agg)
+					return []execinfrapb.ProducerMetadata{*meta}
 				}
-				return []execinfrapb.ProducerMetadata{*meta}
+				return nil
 			},
 		},
 	); err != nil {
@@ -342,36 +342,6 @@ func newStreamIngestionDataProcessor(
 	}
 
 	return sip, nil
-}
-
-func (sip *streamIngestionProcessor) constructTracingAggregatorProducerMeta(
-	ctx context.Context,
-) *execinfrapb.ProducerMetadata {
-	if sip.agg == nil {
-		return nil
-	}
-	aggEvents := &execinfrapb.TracingAggregatorEvents{
-		SQLInstanceID: sip.flowCtx.NodeID.SQLInstanceID(),
-		FlowID:        sip.flowCtx.ID,
-		Events:        make(map[string][]byte),
-	}
-	sip.agg.ForEachAggregatedEvent(func(name string, event bulkutil.TracingAggregatorEvent) {
-		msg, ok := event.(protoutil.Message)
-		if !ok {
-			// This should never happen but if it does skip the aggregated event.
-			log.Warningf(ctx, "event is not a protoutil.Message: %T", event)
-			return
-		}
-		data := make([]byte, msg.Size())
-		if _, err := msg.MarshalTo(data); err != nil {
-			// This should never happen but if it does skip the aggregated event.
-			log.Warningf(ctx, "failed to unmarshal aggregated event: %v", err.Error())
-			return
-		}
-		aggEvents.Events[name] = data
-	})
-
-	return &execinfrapb.ProducerMetadata{AggregatorEvents: aggEvents}
 }
 
 // Start launches a set of goroutines that read from the spans
@@ -397,14 +367,15 @@ func (sip *streamIngestionProcessor) constructTracingAggregatorProducerMeta(
 func (sip *streamIngestionProcessor) Start(ctx context.Context) {
 	ctx = logtags.AddTag(ctx, "job", sip.spec.JobID)
 	log.Infof(ctx, "starting ingest proc")
-	ctx = sip.StartInternal(ctx, streamIngestionProcessorName)
-
-	// Construct an Aggregator to aggregate and render AggregatorEvents emitted in
-	// sips' trace recording.
-	ctx, sip.agg = bulkutil.MakeTracingAggregatorWithSpan(ctx,
-		fmt.Sprintf("%s-aggregator", streamIngestionProcessorName), sip.EvalCtx.Tracer)
+	sip.agg = bulkutil.TracingAggregatorForContext(ctx)
 	sip.aggTimer = timeutil.NewTimer()
-	sip.aggTimer.Reset(15 * time.Second)
+
+	// If the aggregator is nil, we do not want the timer to fire.
+	if sip.agg != nil {
+		sip.aggTimer.Reset(15 * time.Second)
+	}
+
+	ctx = sip.StartInternal(ctx, streamIngestionProcessorName, sip.agg)
 
 	sip.metrics = sip.flowCtx.Cfg.JobRegistry.MetricsStruct().StreamIngest.(*Metrics)
 
@@ -521,7 +492,8 @@ func (sip *streamIngestionProcessor) Next() (rowenc.EncDatumRow, *execinfrapb.Pr
 	case <-sip.aggTimer.C:
 		sip.aggTimer.Read = true
 		sip.aggTimer.Reset(15 * time.Second)
-		return nil, sip.constructTracingAggregatorProducerMeta(sip.Ctx())
+		return nil, bulkutil.ConstructTracingAggregatorProducerMeta(sip.Ctx(),
+			sip.flowCtx.NodeID.SQLInstanceID(), sip.flowCtx.ID, sip.agg)
 	case err := <-sip.errCh:
 		sip.MoveToDraining(err)
 		return nil, sip.DrainHelper()
@@ -584,9 +556,6 @@ func (sip *streamIngestionProcessor) close() {
 	}
 	if sip.maxFlushRateTimer != nil {
 		sip.maxFlushRateTimer.Stop()
-	}
-	if sip.agg != nil {
-		sip.agg.Close()
 	}
 	sip.aggTimer.Stop()
 

--- a/pkg/util/bulk/BUILD.bazel
+++ b/pkg/util/bulk/BUILD.bazel
@@ -10,6 +10,7 @@ go_library(
     importpath = "github.com/cockroachdb/cockroach/pkg/util/bulk",
     visibility = ["//visibility:public"],
     deps = [
+        "//pkg/base",
         "//pkg/clusterversion",
         "//pkg/jobs",
         "//pkg/jobs/jobspb",

--- a/pkg/util/bulk/aggregator_stats.go
+++ b/pkg/util/bulk/aggregator_stats.go
@@ -16,6 +16,7 @@ import (
 	"fmt"
 	"time"
 
+	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/clusterversion"
 	"github.com/cockroachdb/cockroach/pkg/jobs"
 	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
@@ -24,6 +25,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/execinfrapb"
 	"github.com/cockroachdb/cockroach/pkg/sql/isql"
 	"github.com/cockroachdb/cockroach/pkg/sql/protoreflect"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/errors"
 )
@@ -34,6 +36,33 @@ var flushTracingAggregatorFrequency = settings.RegisterDurationSetting(
 	"frequency at which the coordinator node processes and persists tracing aggregator stats to storage",
 	10*time.Minute,
 )
+
+// ConstructTracingAggregatorProducerMeta constructs a ProducerMetadata that
+// contains all the aggregated events stored in the tracing aggregator.
+func ConstructTracingAggregatorProducerMeta(
+	ctx context.Context,
+	sqlInstanceID base.SQLInstanceID,
+	flowID execinfrapb.FlowID,
+	agg *TracingAggregator,
+) *execinfrapb.ProducerMetadata {
+	aggEvents := &execinfrapb.TracingAggregatorEvents{
+		SQLInstanceID: sqlInstanceID,
+		FlowID:        flowID,
+		Events:        make(map[string][]byte),
+	}
+
+	agg.ForEachAggregatedEvent(func(name string, event TracingAggregatorEvent) {
+		if data, err := TracingAggregatorEventToBytes(ctx, event); err != nil {
+			// This should never happen but if it does skip the aggregated event.
+			log.Warningf(ctx, "failed to unmarshal aggregated event: %v", err.Error())
+			return
+		} else {
+			aggEvents.Events[name] = data
+		}
+	})
+
+	return &execinfrapb.ProducerMetadata{AggregatorEvents: aggEvents}
+}
 
 // flushTracingStats persists the following files to the `system.job_info` table
 // for consumption by job observability tools:

--- a/pkg/util/bulk/tracing_aggregator.go
+++ b/pkg/util/bulk/tracing_aggregator.go
@@ -24,8 +24,6 @@ import (
 // TracingAggregator. A TracingAggregatorEvent also implements the tracing.LazyTag interface
 // to render its information on the associated tracing span.
 type TracingAggregatorEvent interface {
-	tracing.LazyTag
-
 	// Identity returns a TracingAggregatorEvent that when combined with another
 	// event returns the other TracingAggregatorEvent unchanged.
 	Identity() TracingAggregatorEvent
@@ -41,8 +39,6 @@ type TracingAggregatorEvent interface {
 // A TracingAggregator can be used to aggregate and render AggregatorEvents that
 // are emitted as part of its tracing spans' recording.
 type TracingAggregator struct {
-	// sp is the tracing span managed by the TracingAggregator.
-	sp *tracing.Span
 	mu struct {
 		syncutil.Mutex
 		// aggregatedEvents is a mapping from the name identifying the
@@ -80,22 +76,14 @@ func (b *TracingAggregator) Notify(event tracing.Structured) tracing.EventConsum
 	eventName := bulkEvent.ProtoName()
 	if _, ok := b.mu.aggregatedEvents[bulkEvent.ProtoName()]; !ok {
 		b.mu.aggregatedEvents[eventName] = bulkEvent.Identity()
-		b.sp.SetLazyTagLocked(eventName, b.mu.aggregatedEvents[eventName])
 	}
 	b.mu.aggregatedEvents[eventName].Combine(bulkEvent)
 	return tracing.EventNotConsumed
 }
 
-// Close is responsible for finishing the TracingAggregator's tracing span.
-//
-// NOTE: it must be called exactly once.
-func (b *TracingAggregator) Close() {
-	b.sp.Finish()
-}
-
 // TracingAggregatorEventToBytes marshals an event into a byte slice.
 func TracingAggregatorEventToBytes(
-	ctx context.Context, event TracingAggregatorEvent,
+	_ context.Context, event TracingAggregatorEvent,
 ) ([]byte, error) {
 	msg, ok := event.(protoutil.Message)
 	if !ok {
@@ -113,27 +101,14 @@ func TracingAggregatorEventToBytes(
 
 var _ tracing.EventListener = &TracingAggregator{}
 
-// MakeTracingAggregatorWithSpan returns an instance of an TracingAggregator along with a
-// newly created child context. The TracingAggregator is registered as a
-// tracing.EventListener on the span associated with newly created context.
-//
-// The TracingAggregator instance is responsible for finishing the returned span, and
-// so the user must call Close().
-func MakeTracingAggregatorWithSpan(
-	ctx context.Context, aggregatorName string, tracer *tracing.Tracer,
-) (context.Context, *TracingAggregator) {
+// TracingAggregatorForContext creates a TracingAggregator if the provided
+// context has a tracing span.
+func TracingAggregatorForContext(ctx context.Context) *TracingAggregator {
 	if tracing.SpanFromContext(ctx) == nil {
-		log.Warningf(ctx, "tracing aggregator %s cannot be created without a tracing span", aggregatorName)
-		return ctx, nil
+		log.Warning(ctx, "tracing aggregator cannot be created without a tracing span")
+		return nil
 	}
 	agg := &TracingAggregator{}
-	aggCtx, aggSpan := tracing.EnsureChildSpan(ctx, tracer, aggregatorName,
-		tracing.WithEventListeners(agg))
-
-	agg.mu.Lock()
-	defer agg.mu.Unlock()
 	agg.mu.aggregatedEvents = make(map[string]TracingAggregatorEvent)
-	agg.sp = aggSpan
-
-	return aggCtx, agg
+	return agg
 }

--- a/pkg/util/bulk/tracing_aggregator_test.go
+++ b/pkg/util/bulk/tracing_aggregator_test.go
@@ -31,12 +31,12 @@ func TestAggregator(t *testing.T) {
 	ctx, root := tr.StartSpanCtx(ctx, "root", tracing.WithRecording(tracingpb.RecordingVerbose))
 	defer root.Finish()
 
-	ctx, agg := bulk.MakeTracingAggregatorWithSpan(ctx, "mockAggregator", tr)
-	aggSp := tracing.SpanFromContext(ctx)
-	defer agg.Close()
+	agg := bulk.TracingAggregatorForContext(ctx)
+	ctx, withListener := tr.StartSpanCtx(ctx, "withListener",
+		tracing.WithEventListeners(agg), tracing.WithParent(root))
+	defer withListener.Finish()
 
-	child := tr.StartSpan("child", tracing.WithParent(root),
-		tracing.WithEventListeners(agg))
+	child := tr.StartSpan("child", tracing.WithParent(withListener))
 	defer child.Finish()
 	child.RecordStructured(&backuppb.ExportStats{
 		NumFiles: 10,
@@ -61,19 +61,19 @@ func TestAggregator(t *testing.T) {
 
 	// We only expect to see the aggregated stats from the local children since we
 	// have not imported the remote children's Recording.
-	exportStats := &backuppb.ExportStats{}
-	exportStatsTag, found := aggSp.GetLazyTag(proto.MessageName(exportStats))
-	require.True(t, found)
-	var es *backuppb.ExportStats
-	var ok bool
-	if es, ok = exportStatsTag.(*backuppb.ExportStats); !ok {
-		t.Fatal("failed to cast LazyTag to expected type")
-	}
-	require.Equal(t, backuppb.ExportStats{
-		NumFiles: 30,
-		DataSize: 30,
-		Duration: 2 * time.Minute,
-	}, *es)
+	agg.ForEachAggregatedEvent(func(name string, event bulk.TracingAggregatorEvent) {
+		require.Equal(t, name, proto.MessageName(&backuppb.ExportStats{}))
+		var es *backuppb.ExportStats
+		var ok bool
+		if es, ok = event.(*backuppb.ExportStats); !ok {
+			t.Fatal("failed to cast event to expected type")
+		}
+		require.Equal(t, backuppb.ExportStats{
+			NumFiles: 30,
+			DataSize: 30,
+			Duration: 2 * time.Minute,
+		}, *es)
+	})
 
 	// Import the remote recording into its parent.
 	rec := remoteChild.FinishAndGetConfiguredRecording()
@@ -81,16 +81,19 @@ func TestAggregator(t *testing.T) {
 
 	// Now, we expect the ExportStats from the remote child to show up in the
 	// aggregator.
-	exportStatsTag, found = aggSp.GetLazyTag(proto.MessageName(exportStats))
-	require.True(t, found)
-	if es, ok = exportStatsTag.(*backuppb.ExportStats); !ok {
-		t.Fatal("failed to cast LazyTag to expected type")
-	}
-	require.Equal(t, backuppb.ExportStats{
-		NumFiles: 60,
-		DataSize: 60,
-		Duration: 3 * time.Minute,
-	}, *es)
+	agg.ForEachAggregatedEvent(func(name string, event bulk.TracingAggregatorEvent) {
+		require.Equal(t, name, proto.MessageName(&backuppb.ExportStats{}))
+		var es *backuppb.ExportStats
+		var ok bool
+		if es, ok = event.(*backuppb.ExportStats); !ok {
+			t.Fatal("failed to cast event to expected type")
+		}
+		require.Equal(t, backuppb.ExportStats{
+			NumFiles: 60,
+			DataSize: 60,
+			Duration: 3 * time.Minute,
+		}, *es)
+	})
 }
 
 func TestIngestionPerformanceStatsAggregation(t *testing.T) {
@@ -123,32 +126,32 @@ func TestIngestionPerformanceStatsAggregation(t *testing.T) {
 		}
 	}
 
-	requireStatsTag := func(aggSp *tracing.Span, expected *bulkpb.IngestionPerformanceStats) {
-		exportStatsTag, found := aggSp.GetLazyTag(expected.ProtoName())
-		require.True(t, found)
-		var actual *bulkpb.IngestionPerformanceStats
-		var ok bool
-		if actual, ok = exportStatsTag.(*bulkpb.IngestionPerformanceStats); !ok {
-			t.Fatal("failed to cast LazyTag to expected type")
-		}
-
-		require.Equal(t, expected, actual)
+	assertAggContainsStats := func(t *testing.T, agg *bulk.TracingAggregator, expected *bulkpb.IngestionPerformanceStats) {
+		agg.ForEachAggregatedEvent(func(name string, event bulk.TracingAggregatorEvent) {
+			require.Equal(t, name, proto.MessageName(expected))
+			var actual *bulkpb.IngestionPerformanceStats
+			var ok bool
+			if actual, ok = event.(*bulkpb.IngestionPerformanceStats); !ok {
+				t.Fatal("failed to cast event to expected type")
+			}
+			require.Equal(t, expected, actual)
+		})
 	}
 
 	// First, start a root tracing span with a tracing aggregator.
 	ctx, root := tr.StartSpanCtx(ctx, "root", tracing.WithRecording(tracingpb.RecordingVerbose))
 	defer root.Finish()
-	ctx, agg := bulk.MakeTracingAggregatorWithSpan(ctx, "mockAggregator", tr)
-	aggSp := tracing.SpanFromContext(ctx)
-	defer agg.Close()
+	agg := bulk.TracingAggregatorForContext(ctx)
+	ctx, withListener := tr.StartSpanCtx(ctx, "withListener",
+		tracing.WithEventListeners(agg), tracing.WithParent(root))
+	defer withListener.Finish()
 
 	// Second, start a child span on the root that also has its own tracing
 	// aggregator.
-	_, child1 := tracing.ChildSpan(ctx, "child1")
-	defer child1.Finish()
-	child1Ctx, child1Agg := bulk.MakeTracingAggregatorWithSpan(ctx, "mockChildAggregator", tr)
-	child1AggSp := tracing.SpanFromContext(child1Ctx)
-	defer child1Agg.Close()
+	child1Agg := bulk.TracingAggregatorForContext(ctx)
+	child1Ctx, child1AggSp := tr.StartSpanCtx(ctx, "child1",
+		tracing.WithEventListeners(child1Agg), tracing.WithParent(withListener))
+	defer child1AggSp.Finish()
 
 	// In addition, start a child span on the first child span.
 	_, child1Child := tracing.ChildSpan(child1Ctx, "child1Child")
@@ -165,6 +168,6 @@ func TestIngestionPerformanceStatsAggregation(t *testing.T) {
 
 	// Verify that the root and child1 aggregators has the expected aggregated
 	// stats.
-	requireStatsTag(child1AggSp, makeEvent(3, map[roachpb.StoreID]time.Duration{1: 203, 2: 202}))
-	requireStatsTag(aggSp, makeEvent(6, map[roachpb.StoreID]time.Duration{1: 203, 2: 405, 3: 303}))
+	assertAggContainsStats(t, child1Agg, makeEvent(3, map[roachpb.StoreID]time.Duration{1: 203, 2: 202}))
+	assertAggContainsStats(t, agg, makeEvent(6, map[roachpb.StoreID]time.Duration{1: 203, 2: 405, 3: 303}))
 }

